File: mobile/ios/Actnet/Sources/Views/Chats/NameGroupView.swift
import SwiftUI

/// Signal-style "Name Group" screen, pushed from the composer's New Group
/// button. Collects the group name, the disappearing-messages timer, and (for
/// an empty group) the hosting server, then creates the group and opens its
/// thread. Group photos are intentionally omitted for now — the core has no
/// avatar param and `docs/30-mobile-ux.md` auto-generates a mosaic icon and
/// defers custom icons at creation.
struct NameGroupView: View {
    @EnvironmentObject var appState: AppState
    @Environment(\.dismiss) private var dismiss

    let members: [ComposeMessageView.Chip]
    let accountId: String
    /// Servers the active identity belongs to; the first is its home server.
    let servers: [ServerInfo]
    /// Called once the group is created — wired by the composer to navigate to
    /// the new thread and dismiss the whole compose sheet.
    let onCreated: (Conversation) -> Void

    @State private var name: String = ""
    @State private var expirySeconds: UInt32 = 0
    @State private var selectedServerId: String = ""
    @State private var creating: Bool = false
    @State private var errorMessage: String?

    /// Home server = the first in the identity's list (matches the rest of the
    /// app, e.g. `AppState`'s `servers.first` server-url derivation).
    private var homeServer: ServerInfo? { servers.first }

    /// Only an empty group may choose a server; with recipients the founder's
    /// home server is used (and a recipient on another homeserver is reached by
    /// federation, not by moving the group). A picker is also pointless with a
    /// single server.
    private var canChooseServer: Bool { members.isEmpty && servers.count > 1 }

    private var resolvedServer: ServerInfo? {
        servers.first(where: { $0.id == selectedServerId }) ?? homeServer
    }

    /// Group name is required — no auto-name from participants.
    private var trimmedName: String {
        name.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    var body: some View {
        Form {
            Section {
                TextField("Group Name (required)", text: $name)
            }

            Section("Server") {
                if canChooseServer {
                    Picker("Server", selection: $selectedServerId) {
                        ForEach(servers) { server in
                            Text(server.displayHost)
                                .tag(server.id)
                                // Non-home creation isn't wired in the core yet
                                // (`create_group` uses the account's pinned
                                // client server), so gate the alternatives.
                                .disabled(server.id != homeServer?.id)
                        }
                    }
                    Text("Creating on another server isn't supported yet.")
                        .font(.caption)
                        .foregroundStyle(.secondary)
                } else if let server = resolvedServer {
                    HStack {
                        Text(server.displayHost)
                        Spacer()
                        Text("Home").font(.caption).foregroundStyle(.secondary)
                    }
                }
            }

            Section {
                DisappearingMessagesPicker(seconds: $expirySeconds)
            }

            Section("Members (\(members.count))") {
                if members.isEmpty {
                    Text("No members yet — you can add people after creating the group.")
                        .font(.footnote)
                        .foregroundStyle(.secondary)
                } else {
                    ForEach(members) { member in
                        HStack(spacing: 10) {
                            ContactAvatar(name: member.label, isBot: isBot(member), size: 32)
                            Text(member.label).lineLimit(1)
                        }
                    }
                }
            }

            if let error = errorMessage {
                Section {
                    Text(error).font(.caption).foregroundStyle(Color.avError)
                }
            }
        }
        .scrollContentBackground(.hidden)
        .background(Color.avPaper)
        .navigationTitle("Name Group")
        .navigationBarTitleDisplayMode(.inline)
        .onAppear {
            if selectedServerId.isEmpty { selectedServerId = homeServer?.id ?? "" }
        }
        .toolbar {
            ToolbarItem(placement: .confirmationAction) {
                Button("Create") { create() }
                    .disabled(creating || trimmedName.isEmpty)
            }
        }
    }

    private func isBot(_ member: ComposeMessageView.Chip) -> Bool {
        appState.isBot(member.did, accountId: accountId)
    }

    private func create() {
        creating = true
        errorMessage = nil
        let title = trimmedName
        let serverUrl = resolvedServer?.id ?? ""
        let recipientDids = members.map(\.did)
        let expiry = expirySeconds
        Task {
            defer { creating = false }
            do {
                let conv = try await appState.createGroupAndOpen(
                    accountId: accountId,
                    serverUrl: serverUrl,
                    title: title,
                    recipientDids: recipientDids,
                    expirySeconds: expiry,
                    firstMessage: nil
                )
                onCreated(conv)
            } catch {
                errorMessage = error.localizedDescription
            }
        }
    }
}

#if DEBUG
#Preview("Group with members") {
    let me = Account(
        id: "did:plc:me",
        displayName: "Me",
        avatarData: nil,
        servers: [ServerInfo(
            id: "https://server.example",
            name: "Example",
            url: URL(string: "https://server.example")!
        )]
    )
    return NavigationStack {
        NameGroupView(
            members: [
                ComposeMessageView.Chip(id: "did:plc:alice", did: "did:plc:alice", displayName: "Alice Rivera"),
                ComposeMessageView.Chip(id: "did:plc:bob", did: "did:plc:bob", displayName: "Bob Chena"),
            ],
            accountId: "did:plc:me",
            servers: me.servers,
            onCreated: { _ in }
        )
        .environmentObject(AppState.preview(accounts: [me]))
    }
}

#Preview("Empty group, multi-server") {
    let me = Account(
        id: "did:plc:me",
        displayName: "Me",
        avatarData: nil,
        servers: [
            ServerInfo(id: "https://a.example", name: "Safe Haven", url: URL(string: "https://a.example")!),
            ServerInfo(id: "https://b.example", name: "Backup Server", url: URL(string: "https://b.example")!),
        ]
    )
    return NavigationStack {
        NameGroupView(
            members: [],
            accountId: "did:plc:me",
            servers: me.servers,
            onCreated: { _ in }
        )
        .environmentObject(AppState.preview(accounts: [me]))
    }
}
#endif

File: mobile/android/app/src/main/kotlin/Views/Chats/NameGroupView.kt
package net.theavalanche.app

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ExposedDropdownMenuBox
import androidx.compose.material3.ExposedDropdownMenuDefaults
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.launch

// ---------------------------------------------------------------------------
// NameGroupView
//
// Signal-style "Name Group" screen, pushed from the composer's New Group
// button. Collects the group name, the disappearing-messages timer, and (for
// an empty group) the hosting server, then creates the group and opens its
// thread.
//
// Mirrors mobile/ios/Actnet/Sources/Views/Chats/NameGroupView.swift.
//
// Navigation: the screen receives lambda callbacks for navigation actions
// (matching the project's lambda-nav pattern). onCreated is called with the
// newly-created Conversation once the group is created.
// ---------------------------------------------------------------------------

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun NameGroupView(
    members: List<RecipientChip>,
    accountId: String,
    /// Servers the active identity belongs to; the first is its home server.
    servers: List<ServerInfo>,
    /// Called once the group is created.
    onCreated: (Conversation) -> Unit = {},
    onDismiss: () -> Unit = {},
    viewModel: AppViewModel,
) {
    // -----------------------------------------------------------------------
    // Local state — mirrors @State vars in the Swift source
    // -----------------------------------------------------------------------

    var name by remember { mutableStateOf("") }
    var expirySeconds by remember { mutableStateOf(0u) }
    var selectedServerId by remember { mutableStateOf("") }
    var creating by remember { mutableStateOf(false) }
    var errorMessage by remember { mutableStateOf<String?>(null) }

    // Seed selectedServerId from the home server on first composition.
    LaunchedEffect(servers) {
        if (selectedServerId.isEmpty()) {
            selectedServerId = servers.firstOrNull()?.id ?: ""
        }
    }

    // -----------------------------------------------------------------------
    // Derived values — mirrors Swift computed vars
    // -----------------------------------------------------------------------

    val homeServer: ServerInfo? = servers.firstOrNull()

    // Only an empty group may choose a server; with recipients the founder's
    // home server is used. A picker is also pointless with a single server.
    val canChooseServer: Boolean = members.isEmpty() && servers.size > 1

    val resolvedServer: ServerInfo? =
        servers.firstOrNull { it.id == selectedServerId } ?: homeServer

    val trimmedName: String = name.trim()

    // -----------------------------------------------------------------------
    // Coroutine scope for the Create action
    // -----------------------------------------------------------------------

    val scope = rememberCoroutineScope()

    fun create() {
        if (creating || trimmedName.isEmpty()) return
        creating = true
        errorMessage = null
        val title = trimmedName
        val serverUrl = resolvedServer?.id ?: ""
        val recipientDids = members.map { it.did }
        val expiry = expirySeconds
        scope.launch {
            try {
                val conv = viewModel.createGroupAndOpen(
                    accountId = accountId,
                    serverUrl = serverUrl,
                    title = title,
                    recipientDids = recipientDids,
                    expirySeconds = expiry,
                    firstMessage = null,
                )
                onCreated(conv)
            } catch (e: Exception) {
                errorMessage = e.localizedMessage ?: e.message ?: "Unknown error"
            } finally {
                creating = false
            }
        }
    }

    // -----------------------------------------------------------------------
    // UI
    // -----------------------------------------------------------------------

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Name Group") },
                navigationIcon = {
                    IconButton(onClick = onDismiss) {
                        Icon(Icons.AutoMirrored.Filled.ArrowBack, contentDescription = "Back")
                    }
                },
                actions = {
                    Button(
                        onClick = { create() },
                        enabled = !creating && trimmedName.isNotEmpty(),
                        colors = ButtonDefaults.buttonColors(
                            containerColor = LocalAvalancheColors.current.brand,
                        ),
                        modifier = Modifier.padding(end = 8.dp),
                    ) {
                        Text("Create")
                    }
                },
                colors = TopAppBarDefaults.topAppBarColors(
                    containerColor = LocalAvalancheColors.current.paper,
                ),
            )
        },
        containerColor = LocalAvalancheColors.current.paper,
    ) { innerPadding ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(innerPadding)
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 16.dp, vertical = 8.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp),
        ) {
            // ----------------------------------------------------------------
            // Group name section
            // ----------------------------------------------------------------

            FormSection {
                OutlinedTextField(
                    value = name,
                    onValueChange = { name = it },
                    label = { Text("Group Name (required)") },
                    singleLine = true,
                    modifier = Modifier.fillMaxWidth(),
                )
            }

            // ----------------------------------------------------------------
            // Server section
            // ----------------------------------------------------------------

            FormSection(label = "Server") {
                if (canChooseServer) {
                    ServerPicker(
                        servers = servers,
                        selectedServerId = selectedServerId,
                        homeServerId = homeServer?.id,
                        onSelectionChange = { selectedServerId = it },
                    )
                    Text(
                        text = "Creating on another server isn't supported yet.",
                        fontSize = 12.sp,
                        color = LocalAvalancheColors.current.muted,
                        modifier = Modifier.padding(top = 4.dp),
                    )
                } else if (resolvedServer != null) {
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        horizontalArrangement = Arrangement.SpaceBetween,
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        Text(resolvedServer.displayHost, color = LocalAvalancheColors.current.ink)
                        Text(
                            text = "Home",
                            fontSize = 12.sp,
                            color = LocalAvalancheColors.current.muted,
                        )
                    }
                }
            }

            // ----------------------------------------------------------------
            // Disappearing messages section
            // ----------------------------------------------------------------

            FormSection {
                DisappearingMessagesPickerView(
                    seconds = expirySeconds,
                    onSecondsChange = { expirySeconds = it },
                )
            }

            // ----------------------------------------------------------------
            // Members section
            // ----------------------------------------------------------------

            FormSection(label = "Members (${members.size})") {
                if (members.isEmpty()) {
                    Text(
                        text = "No members yet — you can add people after creating the group.",
                        fontSize = 13.sp,
                        color = LocalAvalancheColors.current.muted,
                    )
                } else {
                    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                        members.forEachIndexed { index, member ->
                            if (index > 0) {
                                HorizontalDivider(color = LocalAvalancheColors.current.muted.copy(alpha = 0.2f))
                            }
                            Row(
                                verticalAlignment = Alignment.CenterVertically,
                                modifier = Modifier.fillMaxWidth(),
                            ) {
                                val isBot = viewModel.isBot(member.did, accountId = accountId)
                                ContactAvatar(
                                    name = member.label,
                                    isBot = isBot,
                                    size = 32.dp,
                                )
                                Spacer(Modifier.width(10.dp))
                                Text(
                                    text = member.label,
                                    color = LocalAvalancheColors.current.ink,
                                    maxLines = 1,
                                    modifier = Modifier.weight(1f),
                                )
                            }
                        }
                    }
                }
            }

            // ----------------------------------------------------------------
            // Error section
            // ----------------------------------------------------------------

            if (errorMessage != null) {
                FormSection {
                    Text(
                        text = errorMessage!!,
                        fontSize = 12.sp,
                        color = LocalAvalancheColors.current.error,
                    )
                }
            }
        }
    }
}

// ---------------------------------------------------------------------------
// RecipientChip
//
// Mirrors the iOS ComposeMessageView.Chip nested struct. The recipient-chip
// shape is shared across the Chats package via this single declaration;
// ComposeMessageView uses its own recipient model, so there is no duplication.
// ---------------------------------------------------------------------------

data class RecipientChip(
    val id: String,   // == did
    val did: String,
    val displayName: String,
) {
    // User-visible text. Never a raw full DID.
    val label: String
        get() = if (displayName.isEmpty()) shortenDid(did) else displayName
}

// shortenDid lives in ComposeMessageView.kt (same package) — single definition shared across Chats.

// ---------------------------------------------------------------------------
// FormSection — lightweight container that mirrors SwiftUI's Form/Section
// visual grouping (card-style surface with optional label).
// ---------------------------------------------------------------------------

@Composable
private fun FormSection(
    label: String? = null,
    content: @Composable () -> Unit,
) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .background(
                color = LocalAvalancheColors.current.incomingBubble.copy(alpha = 0.5f),
                shape = androidx.compose.foundation.shape.RoundedCornerShape(8.dp),
            )
            .padding(horizontal = 16.dp, vertical = 12.dp),
        verticalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        if (label != null) {
            Text(
                text = label.uppercase(),
                fontSize = 11.sp,
                fontWeight = FontWeight.Medium,
                color = LocalAvalancheColors.current.muted,
                letterSpacing = 0.5.sp,
            )
        }
        content()
    }
}

// ---------------------------------------------------------------------------
// ServerPicker — inline exposed dropdown for the server selector.
// Non-home servers are shown but disabled (matching the iOS .disabled() gate).
// ---------------------------------------------------------------------------

@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun ServerPicker(
    servers: List<ServerInfo>,
    selectedServerId: String,
    homeServerId: String?,
    onSelectionChange: (String) -> Unit,
) {
    var expanded by remember { mutableStateOf(false) }
    val selectedServer = servers.firstOrNull { it.id == selectedServerId }
    val displayValue = selectedServer?.displayHost ?: selectedServerId

    ExposedDropdownMenuBox(
        expanded = expanded,
        onExpandedChange = { expanded = it },
        modifier = Modifier.fillMaxWidth(),
    ) {
        OutlinedTextField(
            value = displayValue,
            onValueChange = {},
            readOnly = true,
            label = { Text("Server") },
            trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded = expanded) },
            colors = ExposedDropdownMenuDefaults.outlinedTextFieldColors(),
            modifier = Modifier
                .fillMaxWidth()
                .menuAnchor(androidx.compose.material3.ExposedDropdownMenuAnchorType.PrimaryNotEditable),
        )

        ExposedDropdownMenu(
            expanded = expanded,
            onDismissRequest = { expanded = false },
        ) {
            servers.forEach { server ->
                val isHome = server.id == homeServerId
                DropdownMenuItem(
                    text = {
                        Row(
                            horizontalArrangement = Arrangement.SpaceBetween,
                            modifier = Modifier.fillMaxWidth(),
                        ) {
                            Text(
                                text = server.displayHost,
                                color = if (isHome) LocalAvalancheColors.current.ink else LocalAvalancheColors.current.muted,
                            )
                            if (isHome) {
                                Text(
                                    text = "Home",
                                    fontSize = 12.sp,
                                    color = LocalAvalancheColors.current.muted,
                                )
                            }
                        }
                    },
                    onClick = {
                        if (isHome) {
                            // Non-home creation isn't wired in the core yet; only
                            // allow selecting the home server (matching iOS .disabled()).
                            onSelectionChange(server.id)
                        }
                        expanded = false
                    },
                    enabled = isHome,
                )
            }
        }
    }
}

// ---------------------------------------------------------------------------
// Previews
// ---------------------------------------------------------------------------

private val previewServers = listOf(
    ServerInfo(
        id = "https://home.example.com",
        name = "Home",
        url = android.net.Uri.parse("https://home.example.com"),
    ),
    ServerInfo(
        id = "https://work.example.com",
        name = "Work",
        url = android.net.Uri.parse("https://work.example.com"),
    ),
)

@Preview(showBackground = true, name = "Group with members")
@Composable
private fun NameGroupWithMembersPreview() {
    AvalancheTheme {
        NameGroupView(
            members = listOf(
                RecipientChip(id = "did:example:bob", did = "did:example:bob", displayName = "Bob"),
                RecipientChip(id = "did:example:carol", did = "did:example:carol", displayName = "Carol"),
            ),
            accountId = "did:example:alice",
            servers = previewServers,
            viewModel = rememberPreviewAppViewModel(),
        )
    }
}

@Preview(showBackground = true, name = "Empty group, multi-server")
@Composable
private fun NameGroupEmptyPreview() {
    AvalancheTheme {
        NameGroupView(
            members = emptyList(),
            accountId = "did:example:alice",
            servers = previewServers,
            viewModel = rememberPreviewAppViewModel(),
        )
    }
}
